### ISSUE TITLE
No double-sided printing (Hp eprint + cloudprint)

### ISSUE THREAD

**Issue by tavinus** [simoncadman/CUPS-Cloud-Print#92]
Hi there. 

First of all, thanks for this awesome piece of software!

I have read all the duplex printing issues from before but none seem to reproduce my problem (they are all closed anyways).

I have an amazon server using this CCP to print to several different google cloud printers. These many printers are connected to google cloudprint in several ways. Some using HP eprint, some using linux server's with armoo's cloudprint connector and some even using windows/chrome. 

We basically use lpr to print from Amazon and we always send options like margins, double-sided-long-edge, etc with the lpr call.

The printers are all working fine, except that some will not print double-sided (duplex).

So, to make it easier I will compare two HP Laserjet P1606dn, one being connected through HP eprint and the other one was installed on a local linux server using hplip (+plugins) and shared to cloudprint with armoo's connector.

The printers show quite differently on the Amazon's CUPS webserver. For example, on the default printer settings there is only a Plex:(Duplex | None) option on the eprint one while we have an option Double-Sided_Printing:( Long_Edge_(Standard) | Short_Edge_(Flip) ) on the other one. The one that does not come from HP eprint has the exact same options as their local linux install counterpart (same as hplip). 

Still, that could be nothing, just the way the generic google cloudprint driver presents itself. Even though that means we can only print duplex long-edge (which is not an issue for us). But when trying to print double-sided, the HP eprint one never works from Amazon. 

So I was thinking this was a HP eprint issue, since it does not print double-sided emails. But Google Cloudprint uses the printer email just to locate the printer. Meaning it does not use the email to print. To be sure my HP eprint printer works double-sided I have opened a text file on my system's chrome and sent it to the printer, with the Duplex option selected and it worked fine. So this means that google cloudprint is capable of sending double-sided prints to my printer, but when sending from amazon + lpr + ccp something is probably not correct.

Using -o sides=two-sided-long-edge with lpr seems to have no effect. Setting Duplex as default also seems to have no effect (it even comes as default).

I did find something weird that may or may not be related.
I tried to run the ccp upgrade script:
```
$ sudo /usr/share/cloudprint-cups/upgrade.py
Fetching list of available ppds...
List retrieved successfully
Updating ATL-PL_HP_P1606dn
ATL-PL_HP_P1606dn not found, you should delete and re-add this printer
Updating ATL-BA_HP_M1212nf
ATL-BA_HP_M1212nf not found, you should delete and re-add this printer
Updating Salvar_no_Google_Drive
Updating ARCO_HP_P1606dn
ARCO_HP_P1606dn not found, you should delete and re-add this printer
Updating ARCO_HP_M1212nf
ARCO_HP_M1212nf not found, you should delete and re-add this printer
Updating APF_HP_M1536dnf
APF_HP_M1536dnf not found, you should delete and re-add this printer
Updating ATL-PL_HP_M1212nf
ATL-PL_HP_M1212nf not found, you should delete and re-add this printer
Updating ARCO_HP_Deskjet2540
ARCO_HP_Deskjet2540 not found, you should delete and re-add this printer
```
So yeah. I removed all printers from CUPS, then removed the account using `deleteaccount.py`and then reinstalled the account and all printers. Which is kind of a pain, because I had to reconfigure defaults for each one again (at least set the paper size to A4).

The problem is that even though I did this, running `$ sudo /usr/share/cloudprint-cups/upgrade.py` still yields the same results, with the `<printer> not found, you should delete and re-add this printer` for all printers.

I am attaching a text file with the results from `$ sudo /usr/share/cloudprint-cups/reportissues.py`. I have just edited the text file, changing the account's email address to "My-Email" (for security reasons).

Another thing that may be useful is this output when trying to get the possible options for the printer:
```
$ lpoptions -h print -p ATL-PL_HP_P1606dn -l
lpoptions: Não é possível obter o arquivo PPD para ATL-PL_HP_P1606dn: Arquivo ou diretório não encontrado
```
It is in portuguese but says:
```
lpoptions: It was not possible to obtain the PPD file for ATL-PL_HP_P1606dn: File or directory not found
```

It seems like the ppd files are quite long names (from the reportissues.py output), it even warns about the name being too long. All printers from cloudprint fail on this lpoptions command, so not sure this helps.

Tomorrow I will test another printer that is connected through HP eprint, but a different model in a different location. This also has duplex and shows exactly the same as my 1606, even though it is another model (they are both using google's generic duplex driver I guess).

Apart from finding a fix, my only other option is to remove the printer from HP eprint (or just google cloud print) and yet again installing it on a local linux server and using armoo's cloudprint connector as the bridge (instead of HP eprint). This method was tested to work with the other 1606.

It really bugs me that double-sided works from chrome to "HP eprint printers" though. Also bugs me that the same printer model prints double-sided from Amazon through armoo's cloudprint connector to linux.

And a final question: What is the recommended way to reinstall a printer? Remove from cups webserver and then just re-add using setupcloudprint.py? It is not really related but I had to remove/add this printer 20 times already and was never sure I was doing it the proper way.

[cups-cloudprint-issues.txt](https://github.com/simoncadman/CUPS-Cloud-Print/files/105934/cups-cloudprint-issues.txt)

**Comment by tavinus** [comment]
Hi again, sorry for double-posting but this is new information and the other post was like 4 days ago.

So, until now there was no way to make the HP LaserJet P1606dn print double-sided. The other duplex printer I have coming through HP eprint prints fine double-sided, using the same settings that fail with the 1606. The weirdest thing is still that I can print duplex from Chrome fine, but when printing from CCP it never works (for this printer).

The firmware was updated and didn't seem to help.

Not sure on what to try next. I guess I will have to disable the HP eprint for this printer and share it through a local linux server. But it really bugs me that this printer has duplex working from Chrome and the other printer has duplex working from from CCP.

With the printer shared from Linux I can even "force" duplex printing by having that as default locally. But having it through HP eprint is a lot more portable and easier to set up.

Any ideas are welcome.
Cheers!

**Comment by tavinus** [comment]
This was a long time ago, but I just found the solution!

The problem was at the printer configuration. and it was the paper's size and type on the tray configuration.

I have opened the printer's IP on Chrome and set the tray's default paper size to "any paper" and the paper type as "any paper" as well. My dialogs are in Portuguese so I am not sure how it is in English. It was set to A4 and common paper before.

Then when the printer received anything it did not consider A4, it would stall and ask to insert a Letter sized paper (for example). Doing this and using the capabilities-fix branch and I have double-sided working. It was a weird bug, sometimes things would go through, sometimes not. Not sure why it sends as letter when we explicitly send it as A4.

So yeah, play with your printer config if you get weird or random errors. Updating firmware may also be important.

This is solved then.
